### PR TITLE
Make SharedDH work for login-non-eldest and provision-paperkey

### PR DESCRIPTION
**Pull request by mlsteele**
Make SDH work for login on second device and provision with paperkey. I'll add the running of existing tests with the feature flag on tomorrow.

cc @patrickxb

### CODE REVIEW

**Comment by patrickxb** [comment]
I think we should be very careful making any changes to signup, login, provisioning, device key gen on the eve of mobile launch.  

The number of devices coming online is (hopefully) going to at least double, and they need to provision in order to do that.  Signups should increase as well.

I understand the importance of this feature, but the shared dh keys aren't going to be used for anything for at least several weeks.  Although there is a feature flag, the code is still changing.  

Delaying a merge into master for a week or two seems prudent to me.

**Comment by maxtaco** [comment]
Let's discuss more in person. The problem with a long-standing branch is it's going to slow us down. we have lots more to build on top of this. It's crucial we get teams done ASAP, I don't want to slow down.  We have to trust the feature flags.  Let's double-check the feature flags are correct.

**Comment by maxtaco** [comment]
@patrickxb I'll add you as a reviewer too.

**Comment by mlsteele** [comment]
So far I've been doing this in a way where the feature flag turns off all the writing and pushing of new data, but some new code still runs for installing the keyring on `G` and not being too careful not to call `Sync` which is new code but effectively no-op. I'll switch to making the feature-flag more of a complete ban on new code. So from now on maybe review this code on two levels. One to make sure it's ok as normal. But at a second level to make sure that when the feature flag is off this shouldn't change execution at all.

Aside from the kex protocol change and sigchain reading, I don't think any new code needs to run while the flag is off. I think it'll work to just be adding new fields which will not be read, and adding functions which completely no-op.

**Comment by patrickxb** [comment]
@mlsteele  So the feature flag changes are going in a different PR?

**Comment by mlsteele** [comment]
This PR failed CI so I'll make sure the parts introduced here are properly flagged at the same time as fixing that. Then another PR to solidify the flagging of already-merged stuff.

File: go/engine/login_provision.go [hunk 1]
```diff
@@ -917,6 +919,13 @@ func (e *loginProvision) ensurePaperKey(ctx *Context) error {
 		}
 	}
 
+	// Load me so that keys will be up to date.
+	var err error
+	e.arg.User, err = libkb.LoadUser(libkb.LoadUserArg{Self: true, UID: e.arg.User.GetUID(), PublicKeyOptional: true, Contextified: libkb.NewContextified(e.G())})
```
Comment (patrickxb) [inline go/engine/login_provision.go]: why is this necessary?

File: go/libkb/globals.go [hunk 2]
```diff
@@ -992,8 +993,8 @@ func (g *GlobalContext) BumpSharedDHKeyring() error {
 	if g.sharedDHKeyring == nil {
 		return makeNew()
 	}
-	eUID, eDid := g.sharedDHKeyring.GetOwner()
-	if eUID.Equal(uid) && eDid.Eq(did) {
+	UID := g.sharedDHKeyring.GetUID()
```
Comment (patrickxb) [inline go/libkb/globals.go]: all lowercase?  `uid`

File: go/engine/device_keygen.go [hunk 3]
```diff
@@ -366,3 +376,45 @@ func (e *DeviceKeygen) device() *libkb.Device {
 		Status:      &s,
 	}
 }
+
+func (e *DeviceKeygen) prepareSDHBoxesFromPaperkey(ctx *Context) ([]keybase1.SharedDHSecretKeyBox, error) {
+	// Assuming this is a paperkey provision.
+
+	sdhk := e.args.SharedDHKeyring
+	if sdhk == nil {
+		debug.PrintStack()
```
Comment (patrickxb) [inline go/engine/device_keygen.go]: do you need this?

File: go/engine/login_provision.go (hunk 1 above)
Comment (mlsteele) [inline go/engine/login_provision.go]: Writing that dissertation made me realize it might not be. I'm going to try something. Meanwhile I'll leave it there so I can remember what's going on. You don't have to read it. :)

---

Our unsuspecting user first signs up on the website. Then they go to login on their first device with SharedDH enabled. They log in via username and password. `LoginProvision` has just provisioned this device using `Run->route->makeEldestDevice`. Now it's time for `ensurePaperKey` which decides to make a new one with `PaperKeyPrimary`. `PaperKeyGen` syncs the `SharedDHKeyring` (to be sure to encrypt all sdh key generations for the new paperkey). The `SharedDHKeyring` needs a `UPAK` to check that the sdh secretboxes came from the right folks. Where does it get this `UPAK`? "Normally" outside of login it would ask the `CachedUPAKLoader`, but the `CachedUPAKLoader` is indisposed*. So we need to pass a `UPAK` to the `SharedDHKeyring.Sync`. This `UPAK` comes from `PaperKeyGenArg.User.ExportToUPAK`. But if that `User` were loaded before `makeEldestDevice` then it wouldn't know about the device key encryption key, and the sync would fail. That is the reason that `User` is re-loaded here. So that the sync `UPAK` knows about the provisioned device key.

* If you use `CachedUPAKLoader` while holding a `LoginContext` it gets most of the way through and then soft-deadlocks on invalidating the `User` cache. I don't actually see a `LoginContext` here...

Comment (mlsteele) [inline go/engine/login_provision.go]: In the same vein as https://github.com/keybase/client/pull/6713, how about leaving this in because doing it another way would be more complicated?

**Comment by mlsteele** [comment]
Lifted the flag up so that very little new code runs when the flag is off. Sorry if all those forces sent notifications.
